v2/tools/generator/internal/codegen/pipeline/create_arm_types.go

```diff
@@ -225,75 +227,106 @@ func (c *armTypeCreator) convertARMPropertyTypeIfNeeded(t astmodel.Type) (astmod
+	}
 
-			if isRequired {
-				// We want to be required but don't need any kubebuidler annotations on this type because it's an ARM type
-				newProp = newProp.MakeRequired().WithKubebuilderRequiredValidation(false)
-			} else {
-				newProp = newProp.MakeOptional()
-			}
+	armPropName := values[0]
+	newProp := astmodel.NewPropertyDefinition(
+		c.idFactory.CreatePropertyName(armPropName, astmodel.Exported),
+		c.idFactory.CreateIdentifier(armPropName, astmodel.NotExported),
+		astmodel.StringType)
+
+	if isRequired {
+		// We want to be required but don't need any kubebuidler annotations on this type because it's an ARM type
+		newProp = newProp.MakeRequired().WithKubebuilderRequiredValidation(false)
+	} else {
+		newProp = newProp.MakeOptional()
+	}
 
-			result = result.WithProperty(newProp)
-		} else {
-			newType, err := c.convertARMPropertyTypeIfNeeded(prop.PropertyType())
+	return newProp, nil
+}
+
+func (c *armTypeCreator) createARMProperty(prop *astmodel.PropertyDefinition, _ bool) (*astmodel.PropertyDefinition, error) {
+	newType, err := c.createARMTypeIfNeeded(prop.PropertyType())
 
+	if err != nil {
+		return nil, err
+	}
+	return prop.WithType(newType), nil
+}
+
+// convertObjectPropertiesForARM returns the given object type with
+// any properties updated that need to be changed for ARM
+func (c *armTypeCreator) convertObjectPropertiesForARM(t *astmodel.ObjectType, isSpecType bool) (*astmodel.ObjectType, error) {
+	propertyHandlers := []armPropertyTypeConversionHandler{
+		c.createARMNameProperty,
+		c.createResourceReferenceProperty,
+		c.createARMProperty,
+	}
+
+	result := t.WithoutProperties()
+	var errs []error
+	for _, prop := range t.Properties() {
+		for _, handler := range propertyHandlers {
+			newProp, err := handler(prop, isSpecType)
 			if err != nil {
 				errs = append(errs, err)
-			} else {
-				result = result.WithProperty(prop.WithType(newType))
+				break // Stop calling handlers and proceed to the next property
+			}
+
+			if newProp != nil {
+				result = result.WithProperty(newProp)
+				break
 			}
 		}
 	}
 
+	embeddedPropertyHandlers := []armPropertyTypeConversionHandler{
+		c.createARMProperty,
+	}
+
 	// Also convert embedded properties if there are any
 	result = result.WithoutEmbeddedProperties() // Clear them out first so we're starting with a clean slate
 	for _, prop := range t.EmbeddedProperties() {
-		newType, err := c.convertARMPropertyTypeIfNeeded(prop.PropertyType())
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
-		result, err = result.WithEmbeddedProperty(prop.WithType(newType))
-		if err != nil {
-			errs = append(errs, err)
-			continue
+		for _, handler := range embeddedPropertyHandlers {
+			newProp, err := handler(prop, isSpecType)
+			if err != nil {
+				errs = append(errs, err)
+				break // Stop calling handlers and proceed to the next property
+			}
+
+			if newProp != nil {
+				result, err = result.WithEmbeddedProperty(newProp)
```

Comment:
Wouldn't it be more efficient to update the property outside the loop, after all the property handlers have made their change?

Reply:
Actually, every handler is called with `prop` (initialized on line 315) - but that's never updated from `newProp`, so if there are multiple handlers the results of the earlier ones are discarded and overwritten.

Reply:
Hmm, that doesn't make it a ton clearer either at least for me.

Maybe I should just put a comment on the break inside `newProp != nil` that says:
`// Once the first property has been matched, stop.`

We have similar code to this elsewhere and that's the pattern it follows - for example see [`renamer.go` line 177](https://github.com/Azure/azure-service-operator/blob/1f708876c2d9d906c93c788781e2a4ea739a92e6/v2/tools/generator/internal/codegen/embeddedresources/renamer.go#L177), or [`property_conversions.go` line 158](https://github.com/Azure/azure-service-operator/blob/main/v2/tools/generator/internal/conversions/property_conversions.go#L163) (which has a return rather than a break but otherwise is similar).

Reply:
I've put the comments, see what you think

Reply:
Works for me. :+1: